app/models/merged_search_service.rb

```diff
@@ -0,0 +1,241 @@
+  #
+  # WARNING: exceptions raised inside these threads will not automatically
+  # propagate to the caller; callers/tests should account for this.
+  #
+  # @param offset [Integer] api offset to request
+  # @param per_page [Integer] number of items to request
+  # @return [Array<Hash,Hash>] [primo_response, timdex_response]
+  def parallel_fetch(offset:, per_page:)
+    primo = nil
+    timdex = nil
+    threads = []
+    threads << Thread.new { primo = @primo_fetcher.call(offset: offset, per_page: per_page, query: @enhanced_query) }
+    threads << Thread.new { timdex = @timdex_fetcher.call(offset: offset, per_page: per_page, query: @enhanced_query) }
+    threads.each(&:join)
+    [primo, timdex]
+  end
+
+  # Compute API offsets from the paginator and fetch the page-sized chunks
+  # required to assemble the merged page.
+  #
+  # @param paginator [MergedSearchPaginator]
+  # @return [Array<Hash,Hash>] [primo_data, timdex_data]
+  def fetch_all_tab_page_chunks(paginator)
+    merge_plan = paginator.merge_plan
+    primo_count = merge_plan.count(:primo)
+    timdex_count = merge_plan.count(:timdex)
+    primo_offset, timdex_offset = paginator.api_offsets
+
+    primo_thread = if primo_count > 0
+                     Thread.new do
+                       @primo_fetcher.call(offset: primo_offset, per_page: primo_count, query: @enhanced_query)
+                     end
+                   end
+    timdex_thread = if timdex_count > 0
+                      Thread.new do
+                        @timdex_fetcher.call(offset: timdex_offset, per_page: timdex_count, query: @enhanced_query)
+                      end
+                    end
+
+    primo_data = if primo_thread
+                   primo_thread.value
+                 else
+                   { results: [], errors: nil, hits: paginator.primo_total,
+                     show_continuation: false }
+                 end
+    timdex_data = timdex_thread ? timdex_thread.value : { results: [], errors: nil, hits: paginator.timdex_total }
+
+    [primo_data, timdex_data]
+  end
+
+  # Assemble the final hash returned to the controller for rendering.
+  #
+  # @param paginator [MergedSearchPaginator]
+  # @param primo_data [Hash] response from Primo fetcher
+  # @param timdex_data [Hash] response from TIMDEX fetcher
+  # @param current_page [Integer]
+  # @param per_page [Integer]
+  # @param deeper [Boolean] whether this was a deeper-page flow
+  # @return [Hash] response with :results, :errors, :pagination, :show_primo_continuation
+  def assemble_all_tab_result(paginator, primo_data, timdex_data, current_page, per_page, deeper: false)
+    primo_total = primo_data[:hits] || 0
+    timdex_total = timdex_data[:hits] || 0
+
+    merged = paginator.merge_results(primo_data[:results] || [], timdex_data[:results] || [])
+    errors = combine_errors(primo_data[:errors], timdex_data[:errors])
+    pagination = Analyzer.new(@enhanced_query, timdex_total, :all, primo_total).pagination
```

Comment:
I'm not sure if this is a bug that can happen in this application, or just in the console.

But here's a sequence of console commands that lead to an unexpected result for this pagination value (tldr; it uses the per_page for the app whereas all of these methods allow you to pass any value you want... maybe they shouldn't accept values and just use the per_page of the app to avoid having someone override something at some point and have pagination blow up?)

```
params = {q: 'hello'}
q = Enhanger.new(params)
s = MergedSearchService.new(enhanced_query: q, active_tab: 'all')
s.fetch(page: 1, per_page: 1)
```

Pagination will then be:
```
pagination: {hits: 162397, per_page: 20, start: 1, end: 20, next: 2}
```

Which has the correct offsets and next, but the wrong per_page.